### ISSUE TITLE
[GENERAL SUPPORT]: Spreading an NAS over multiple accounts

### ISSUE THREAD

**Issue by CCranney** [facebook/Ax#3609]
### Question

Hi,

I'm designing an NAS that relies on this package. This NAS will be particularly intensive, so I am trying to find ways to spread it across multiple lab member computers/accounts. I sketched out an outline of a possible experiment, but there are key features I don't know are possible using Ax. I thought I'd reach out to see if you know of a way to implement this approach.

My code is largely based on the article for performing [Multi-Objective NAS with Ax](https://pytorch.org/tutorials/intermediate/ax_multiobjective_nas_tutorial.html). The tutorial saves selected NAS-determined metrics (and resulting losses) using Tensorboard. In a dummy example with num_trials set to 20, the first 5 trials are generated using the Sobol strategy, the remaining 15 using BO_MIXED. To rephrase, there are 5 completely random trials, followed by 15 specializing or refining trials.

My idea was to perform many random trials across multiple devices, gather up the results, then run an NAS that runs only BO_MIXED generation strategies on a single device (the "real" NAS) using the results generated from the random trials. However, I am struggling to find a way to adjust the TorchXRunner, GenerationStrategy, and Scheduler classes to create this multi-run NAS experimental design. I would need to set up a generator and scheduler that can reference pre-generated results in future generations/scheduling from the very beginning.

The closest method I am finding is in your [scheduler tutorial](https://ax.dev/docs/tutorials/scheduler/#6-leveraging-sql-storage-and-experiment-resumption), where you indicate that experiments can be interrupted and then resumed if one uses an sqlite database. This demonstrates that a scheduler can reference prior data when it starts. However, it's not quite the same - it's not run from a manually-curated combination of prior experiment results.

Do you see a way I could perform this kind of experiment? I'm very open to drastic rewrites of my code, but am not sure if it's even feasible with Ax.

For reference, I'm attaching a jupyter notebook with my own NAS tutorial I wrote up for lab members. It's not my final use case, but the general workflow follows what I am doing.

[NAS_tutorial.tar.gz](https://github.com/user-attachments/files/19636548/NAS_tutorial.tar.gz)

### Please provide any relevant code snippet if applicable.

```shell

```

### Code of Conduct

- [x] I agree to follow this Ax's Code of Conduct

**Comment by bernardbeckerman** [comment]
Hi @CCranney,

You're right that Ax can "warm-start" a later experiment from an earlier one, and that combining multiple pre-existing experiments is currently something that's not well supported. Optimizations also tend to perform better if run via one central process (even our "random search" is not fully random, but a pseudo-random sequence designed to be more exploratory than pure random search). A few questions/options, since I'm not sure what is possible given your setup:

1. Is it possible for you to run one central process that launches jobs on separate machines and reads their results? If so you can implement a Runner and Metric to automate this process and run a completely automated optimization via the Scheduler.
2. If not, how many trials are you thinking of performing, and how often will you run this optimization? Depending on your needs, it may be easier to generate parameterizations, launch trials, and log results manually via the [Client](https://ax.dev/docs/next/tutorials/ask_tell/). 
3. It's technically possible to run a single experiment on multiple machines using multiple Clients backed by a central database, but this would be the least well supported option, since you'd have to manage concurrency issues and keep all experiment instances up-to-date with each other yourself.

Please let me know if any of these options work for you.

**Comment by CCranney** [comment]
Thank you so much for your response! That is good to know that Sobol is not fully random, it is certainly helpful in planning this out! While it may not be fully random, I would assume running the same job with different seeds would still have different iterative "random" experimentations, correct? If you were to run 3 jobs (10 Sobol-generated tasks each) concurrently, would you expect each of those three to mirror the others, or would they identify different optimization "paths?"

1. I'm afraid this would not immediately work - if push comes to shove, it looks like I can run the script such that different scripts are submitted as batch jobs (SLURM) to parallelize them, but access to computational resources as an individual are limited in my organization, so it's still not ideal. The hope is that I could have multiple lab members run the same script for the randomized portion, then combine the results after they have completed for the larger, more refined NAS search task (which I assume definitely needs to be run from the same script).
2. This is still up for debate - I've created a full search space for consideration, but will be reducing it depending on how many computational resources I can utilize. Each iteration of the model I hope to train will likely take at least a day, but am hoping for about 30 initial, random iterations, followed by as many iterations as I can manage (hopefully upwards of ~50, but I don't know if I can swing that).
3. Yes, this sounds less than ideal - I was more hoping that, if the randomized results are all stored in a single directory, an NAS that skips the randomization step could use the outputs stored in that model to supplement them. If that's not possible, or not recommended, I'll just go the slow way.

**Comment by bernardbeckerman** [comment]
> While it may not be fully random, I would assume running the same job with different seeds would still have different iterative "random" experimentations, correct?

Yes, they will be different. The advantage of the pseudo-random Sobol generator we use for our initialization batch is that random sequences have some chance of re-suggesting similar points, whereas Sobol is more "spread out" in a few important ways. If you use 3 different processes, the random initialization will be slightly less trial-efficient.

> If you were to run 3 jobs (10 Sobol-generated tasks each) concurrently, would you expect each of those three to mirror the others, or would they identify different optimization "paths?"

The random initializations would be independent for each job, and the optimization would proceed in slightly different ways for each tuning job. Generally with enough trials, they should converge on the same point in your search space, so your intuition to perform the last batch from a central process is a good one.

Perhaps some semi-automated version of (2) could work, for instance:
1. Generate 30 sobol trials from a central AxService instance.
2. Loop over 10 apiece on each of 3 machines and collect the results.
3. Load the results in to your central AxService instance.
4. Proceed with the serial model-based experimentation on a single machine (you can also do this in batch mode as well, e.g., suggest 6 trials, run 2 per machine, load into AxService, and repeat).

Does this make sense?

**Comment by bernardbeckerman** [comment]
You can follow the Service API's tutorial for this, specifically, "4. Run optimization loop" ([link](https://ax.dev/docs/tutorials/gpei_hartmann_service/#4-run-optimization-loop)), except in this case you would split up the contents of the loop into trial generation, i.e., running the following line n times,

```
parameterization, trial_index = ax_service.get_next_trial()
```

and then result logging, i.e., compile results for each trial into a `results_dict` mapping metric name to value (similar to the return value of the evaluate function in the tutorial), then run

```
ax_client.complete_trial(trial_index=trial_index, raw_data=results_dict)
```

Of course, you'd have to keep track of which trial_index corresponds to which result_dict yourself, but as long as you do that, you should be fine to use this setup.

Lmk if this all makes sense and will work for your use-case.

**Comment by CCranney** [comment]
That does make sense, thank you! I think I have enough to work with - I'll close this issue for now, and come back if I run into any problems. Thank you again so much for walking me through the logistics of my unique use case!
